library/management/chat/src/main/java/org/quiltmc/qsl/chat/mixin/client/ClientPlayNetworkHandlerMixin.java

```diff
@@ -0,0 +1,172 @@
+		ProfileIndependentS2CMessage message = new ProfileIndependentS2CMessage(client.player, true, packet);
+		return (ProfileIndependentMessageS2CPacket)QuiltChatEvents.MODIFY.invoke(message, message).serialized();
+	}
+
+	@Inject(method = "onProfileIndependentMessage", at = @At(value = "FIELD", target = "Lnet/minecraft/client/network/ClientPlayNetworkHandler;client:Lnet/minecraft/client/MinecraftClient;", shift = At.Shift.BEFORE, ordinal = 1), cancellable = true)
+	public void quilt$cancelInboundProfileIndependentMessage(ProfileIndependentMessageS2CPacket packet, CallbackInfo ci) {
+		ProfileIndependentS2CMessage message = new ProfileIndependentS2CMessage(client.player, true, packet);
+		if (QuiltChatEvents.CANCEL.invoke(message) == Boolean.TRUE) {
+			ci.cancel();
+		}
+	}
+
+	@Inject(method = "onProfileIndependentMessage", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/MinecraftClient;m_xlacueca()Lnet/minecraft/client/gui/ClientChatListener;"))
+	public void quilt$beforeInboundProfileIndependentMessage(ProfileIndependentMessageS2CPacket packet, CallbackInfo ci) {
+		ProfileIndependentS2CMessage message = new ProfileIndependentS2CMessage(client.player, true, packet);
+		QuiltChatEvents.BEFORE_PROCESS.invoke(message);
+	}
+
+	@Inject(method = "onProfileIndependentMessage", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/gui/ClientChatListener;m_jytgvbam(Lnet/minecraft/text/Text;Lnet/minecraft/network/message/MessageType$Parameters;)V", shift = At.Shift.AFTER))
+	public void quilt$afterInboundProfileIndependentMessage(ProfileIndependentMessageS2CPacket packet, CallbackInfo ci) {
+		ProfileIndependentS2CMessage message = new ProfileIndependentS2CMessage(client.player, true, packet);
+		QuiltChatEvents.AFTER_PROCESS.invoke(message);
+	}
+
+	@ModifyVariable(method = "m_fzlgisyq", at = @At("HEAD"), argsOnly = true)
+	public String quilt$modifyOutboundRawChatMessage(String string) {
+		// Not sure *why* this would be null but IDEA is complaining, so, safety first?
+		if (client.player == null) {
+			return string;
+		}
+
+		RawChatC2SMessage message = new RawChatC2SMessage(client.player, true, string);
+		return ((RawChatC2SMessage)QuiltChatEvents.MODIFY.invoke(message, message)).serialized();
+	}
+
+	@Inject(method = "m_fzlgisyq", at = @At(value = "HEAD"), cancellable = true)
+	public void quilt$cancelAndBeforeOutboundRawChatMessage(String string, CallbackInfo ci) {
+		if (client.player == null) {
+			return;
+		}
+
+		RawChatC2SMessage message = new RawChatC2SMessage(client.player, true, string);
+		if (QuiltChatEvents.CANCEL.invoke(message)) {
+			ci.cancel();
+		} else {
+			QuiltChatEvents.BEFORE_PROCESS.invoke(message);
+		}
+	}
+
+	@Inject(method = "m_fzlgisyq", at = @At(value = "TAIL"))
+	public void quilt$afterOutboundRawChatMessage(String string, CallbackInfo ci) {
+		if (client.player == null) {
+			return;
+		}
```

Comment:
Hm, maybe?
```suggestion
		if (client.player == null) return;
```
I don't know if others would agree, but I feel like it would be cleaner to inline it here